### DISCUSSION TITLE
Why Spring batch allows job without parameters to restart even if the job already completed successfullly?

### DISCUSSION THREAD

**Discussion by ELMORABITYounes** [spring-projects/spring-batch#4694] [Q&A]
Right now even if a job was completed successfuly, spring batch allow It to be restarted if It contains no identifying params as shown here:

```				
if (!identifyingJobParameters.isEmpty()                                                        
		&& (status == BatchStatus.COMPLETED || status == BatchStatus.ABANDONED)) {            
	throw new JobInstanceAlreadyCompleteException(                                             
			"A job instance already exists and is complete for identifying parameters="       
					+ identifyingJobParameters + ".  If you want to run this job again, "    
					+ "change the parameters.");                                             
}                                                                                              
```

I am wondering why is that done like this? I mean if the job already completed why It does not throw JobInstanceAlreadyCompleteException? Shouldn't second job instance without param considered the same and hence not allow It to be restarted if already succeeded?

**Comment by fmbenhassine** [comment]
Is that really the case? This would be a bug. I did not test that (yet), but if you provide an example I will debug the case. No matter what the user provides, if the set of identifying job parameters ends up being the same (ie same hash), then the same job instance should not be restarted if it was completed in the previous run.

**Reply by ELMORABITYounes** [reply]
Hello @fmbenhassine,

Thank you for your response!

You can find an example for reproducing the issue [here](https://github.com/ELMORABITYounes/spring-batch-completed-job-rerun-issue).

In that repo I created a generic api for running async jobs by the given job name, and what I am expecting is that when I run the same job twice I should get a `JobInstanceAlreadyCompleteException` exception if the job already completed successfully, but as I mentioned, if I don't provide any identifying parameter, the job always runs successfully even though it has already been completed.

In the branch `job-run-with-identifiers` you can see that when adding a default idenitifying parameter, the result is as expected and the expected exception is thrown.

As you can see in the above code snippet I shared previously It seems that this case was addressed specifically; that is why I thought maybe this is an intentional behavior and maybe there is some specific reason behind that.

**Reply by fmbenhassine** [reply]
Thank you for the example. It was not minimal (it includes web and docker/postgres to run, but that is not an issue, thank you for your time!), so I tried to isolate the case in this example:

```java
package org.springframework.batch.samples.helloworld;

import org.springframework.batch.core.Job;
import org.springframework.batch.core.JobParameters;
import org.springframework.batch.core.JobParametersBuilder;
import org.springframework.batch.core.Step;
import org.springframework.batch.core.configuration.annotation.EnableBatchProcessing;
import org.springframework.batch.core.job.builder.JobBuilder;
import org.springframework.batch.core.launch.JobLauncher;
import org.springframework.batch.core.repository.JobRepository;
import org.springframework.batch.core.step.builder.StepBuilder;
import org.springframework.batch.repeat.RepeatStatus;
import org.springframework.context.ApplicationContext;
import org.springframework.context.annotation.AnnotationConfigApplicationContext;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.jdbc.datasource.embedded.EmbeddedDatabaseBuilder;
import org.springframework.jdbc.support.JdbcTransactionManager;

import javax.sql.DataSource;

@Configuration
@EnableBatchProcessing
public class HelloWorldJobConfiguration {

	public static void main(String[] args) throws Exception {
		ApplicationContext context = new AnnotationConfigApplicationContext(HelloWorldJobConfiguration.class);
		JobLauncher jobLauncher = (JobLauncher) context.getBean("jobLauncher");
		Job job = (Job) context.getBean("job");
		JobParameters jobParameters1 = new JobParametersBuilder().addString("name", "foo", false).toJobParameters();
		JobParameters jobParameters2 = new JobParametersBuilder().addString("name", "bar", false).toJobParameters();
		jobLauncher.run(job, jobParameters1);
		jobLauncher.run(job, jobParameters2); // expected: JobInstanceAlreadyCompleteException
	}

	@Bean
	public Job job(JobRepository jobRepository, Step step) {
		return new JobBuilder("job", jobRepository).start(step).build();
	}

	@Bean
	public Step step(JobRepository jobRepository, JdbcTransactionManager transactionManager) {
		return new StepBuilder("step", jobRepository).tasklet((contribution, chunkContext) -> {
			System.out.println("Hello world!");
			return RepeatStatus.FINISHED;
		}, transactionManager).build();
	}

	// infra beans

	@Bean
	public DataSource dataSource() {
		return new EmbeddedDatabaseBuilder()
				.addScript("/org/springframework/batch/core/schema-hsqldb.sql")
				.build();
	}

	@Bean
	public JdbcTransactionManager transactionManager(DataSource dataSource) {
		return new JdbcTransactionManager(dataSource);
	}

}
```

while the default job key generator works as expected (it gives the same hash for the same input, ie an empty identifying job parameters set), spring batch still considers this as a different job instance and runs it, which should not be the case.

I will turn this discussion into an issue and plan it for the next release. Thank you @ELMORABITYounes !

---

Just FTR,  the default job key generator is working as expected (the following tests pass with 5.2.1):

```java
// to add in org.springframework.batch.core.DefaultJobKeyGeneratorTests

	@Test
	public void testCreateJobKeyForEmptyParameters() {
		JobParameters jobParameters1 = new JobParameters();
		JobParameters jobParameters2 = new JobParameters();
		String key1 = jobKeyGenerator.generateKey(jobParameters1);
		String key2 = jobKeyGenerator.generateKey(jobParameters2);
		assertEquals(key1, key2);
	}

	@Test
	public void testCreateJobKeyForEmptyParametersAndNonIdentifying() {
		JobParameters jobParameters1 = new JobParameters();
		JobParameters jobParameters2 = new JobParametersBuilder()
				.addString("name", "foo", false)
				.toJobParameters();
		String key1 = jobKeyGenerator.generateKey(jobParameters1);
		String key2 = jobKeyGenerator.generateKey(jobParameters2);
		assertEquals(key1, key2);
	}
```

**Comment by fmbenhassine** [comment]
Closing this discussion as I turned it into an issue: https://github.com/spring-projects/spring-batch/issues/4755
